File: go/proxx/proxx.go
package proxx

import (
	"bufio"
	"fmt"
	"math"
	"math/rand"
	"os"
	"strconv"
	"strings"
	"time"
)

type ProxxGame struct {
	Width        int
	Height       int
	NumHoles     int
	Board        [][]int
	VisibleBoard [][]string
}

func (g *ProxxGame) getBoardSize() {
	scanner := bufio.NewScanner(os.Stdin)
	for {
		fmt.Print("Enter the width of the board: ")
		scanner.Scan()
		width, err := strconv.Atoi(scanner.Text())
		if err == nil {
			g.Width = width
			break
		}
		fmt.Println("Invalid input. Please enter a valid integer.")
	}

	for {
		fmt.Print("Enter the height of the board: ")
		scanner.Scan()
		height, err := strconv.Atoi(scanner.Text())
		if err == nil {
			g.Height = height
			break
		}
		fmt.Println("Invalid input. Please enter a valid integer.")
	}
}

func (g *ProxxGame) getNumHoles() {
	scanner := bufio.NewScanner(os.Stdin)
	for {
		fmt.Print("Enter the number of holes: ")
		scanner.Scan()
		numHoles, err := strconv.Atoi(scanner.Text())
		if err == nil {
			if numHoles >= g.Width*g.Height {
				fmt.Println("Invalid input. Number of holes cannot exceed the total number of cells.")
			} else {
				g.NumHoles = numHoles
				break
			}
		} else {
			fmt.Println("Invalid input. Please enter a valid integer.")
		}
	}
}

func (g *ProxxGame) createBoard() {
	g.Board = make([][]int, g.Height)
	g.VisibleBoard = make([][]string, g.Height)

	for i := 0; i < g.Height; i++ {
		g.Board[i] = make([]int, g.Width)
		g.VisibleBoard[i] = make([]string, g.Width)
		for j := 0; j < g.Width; j++ {
			g.VisibleBoard[i][j] = "-"
		}
	}
}

func (g *ProxxGame) placeHoles() {
	rand.Seed(time.Now().UnixNano())

	locations := rand.Perm(g.Width * g.Height)[:g.NumHoles]
	for _, location := range locations {
		row := location / g.Width
		col := location % g.Width
		g.Board[row][col] = -1
	}
}

func (g *ProxxGame) calculateNumbers() {
	for row := 0; row < g.Height; row++ {
		for col := 0; col < g.Width; col++ {
			if g.Board[row][col] != -1 {
				count := 0
				for r := int(math.Max(0, float64(row)-1)); r <= int(math.Min(float64(row)+1, float64(g.Height)-1)); r++ {
					for c := int(math.Max(0, float64(col)-1)); c <= int(math.Min(float64(col)+1, float64(g.Width)-1)); c++ {
						if g.Board[r][c] == -1 {
							count++
						}
					}
				}
				g.Board[row][col] = count
			}
		}
	}
}

func (g *ProxxGame) printBoard() {
	for _, row := range g.VisibleBoard {
		fmt.Println(strings.Join(row, " "))
	}
}

func (g *ProxxGame) revealCell(row, col int) bool {
	if g.Board[row][col] == -1 {
		g.VisibleBoard[row][col] = "H"
		return false
	}

	g.VisibleBoard[row][col] = strconv.Itoa(g.Board[row][col])

	if g.Board[row][col] == 0 {
		for r := int(math.Max(0, float64(row)-1)); r <= int(math.Min(float64(row)+1, float64(g.Height-1))); r++ {
			for c := int(math.Max(0, float64(col)-1)); c <= int(math.Min(float64(col)+1, float64(g.Width-1))); c++ {
				if g.VisibleBoard[r][c] == "-" {
					g.revealCell(r, c)
				}
			}
		}
	}
	return true
}

func (g *ProxxGame) checkWin() bool {
	for row := 0; row < g.Height; row++ {
		for col := 0; col < g.Width; col++ {
			if g.Board[row][col] != -1 && g.VisibleBoard[row][col] == "-" {
				return false
			}
		}
	}
	return true
}

func (g *ProxxGame) Play() {
	g.getBoardSize()
	g.getNumHoles()
	g.createBoard()
	g.placeHoles()
	g.calculateNumbers()

	fmt.Println("Initial board:")
	g.printBoard()

	scanner := bufio.NewScanner(os.Stdin)
	for {
		fmt.Print("Enter row (0 to ", g.Height-1, "): ")
		scanner.Scan()
		row, err := strconv.Atoi(scanner.Text())
		if err != nil {
			fmt.Println("Invalid input. Please enter a valid integer.")
			continue
		}

		fmt.Print("Enter column (0 to ", g.Width-1, "): ")
		scanner.Scan()
		col, err := strconv.Atoi(scanner.Text())
		if err != nil {
			fmt.Println("Invalid input. Please enter a valid integer.")
			continue
		}

		if row < 0 || row >= g.Height || col < 0 || col >= g.Width {
			fmt.Println("Invalid move. Try again.")
			continue
		}

		if !g.revealCell(row, col) {
			fmt.Println("Game over! You hit a hole.")
			break
		}

		if g.checkWin() {
			fmt.Println("Congratulations! You won the game!")
			break
		}

		fmt.Println("--------------------")
		g.printBoard()
	}

	fmt.Println("Final board:")
	g.printBoard()
}

File: python/src/proxx.py
import random
from typing import List

class ProxxGame:
    def __init__(self):
        """Initialize the Proxx game."""
        self.width: int = 0
        self.height: int = 0
        self.num_holes: int = 0
        self.board: List[List[int]] = []
        self.visible_board: List[List[str]] = []

    def get_board_size(self) -> None:
        """Prompt the player to enter the width and height of the board."""
        while True:
            try:
                self.width = int(input("Enter the width of the board: "))
                self.height = int(input("Enter the height of the board: "))
                break
            except ValueError:
                print("Invalid input. Please enter a valid integer.")

    def get_num_holes(self) -> None:
        """Prompt the player to enter the number of holes."""
        while True:
            try:
                self.num_holes = int(input("Enter the number of holes: "))
                if self.num_holes >= self.width * self.height:
                    print("Invalid input. Number of holes cannot exceed the total number of cells.")
                else:
                    break
            except ValueError:
                print("Invalid input. Please enter a valid integer.")

    def create_board(self) -> None:
        """Create the game board and the visible board."""
        self.board = [[0 for _ in range(self.width)] for _ in range(self.height)]
        self.visible_board = [["-" for _ in range(self.width)] for _ in range(self.height)]

    def place_holes(self) -> None:
        """Randomly place the holes on the game board."""
        locations = random.sample(range(self.width * self.height), self.num_holes)
        for location in locations:
            row = location // self.width
            col = location % self.width
            self.board[row][col] = -1

    def calculate_numbers(self) -> None:
        """Calculate the numbers indicating the neighboring holes for each cell."""
        for row in range(self.height):
            for col in range(self.width):
                if self.board[row][col] != -1:
                    count = 0
                    for r in range(max(0, row - 1), min(row + 2, self.height)):
                        for c in range(max(0, col - 1), min(col + 2, self.width)):
                            if self.board[r][c] == -1:
                                count += 1
                    self.board[row][col] = count

    def print_board(self) -> None:
        """Print the visible board."""
        for row in self.visible_board:
            print(" ".join(str(cell) for cell in row))

    def reveal_cell(self, row: int, col: int) -> bool:
        """
        Reveal the selected cell.

        Args:
            row (int): The row index of the cell.
            col (int): The column index of the cell.

        Returns:
            bool: True if the cell is successfully revealed, False if it contains a hole.

        """
        if self.board[row][col] == -1:
            self.visible_board[row][col] = "H"
            return False
        
        self.visible_board[row][col] = str(self.board[row][col])
        
        if self.board[row][col] == 0:
            for r in range(max(0, row - 1), min(row + 2, self.height)):
                for c in range(max(0, col - 1), min(col + 2, self.width)):
                    if self.visible_board[r][c] == "-":
                        self.reveal_cell(r, c)
        return True

    def check_win(self) -> bool:
        """
        Check if the player has won the game.

        Returns:
            bool: True if the player has won, False otherwise.

        """
        for row in range(self.height):
            for col in range(self.width):
                if self.board[row][col] != -1 and self.visible_board[row][col] == "-":
                    return False
        return True

    def play(self) -> None:
        """Play the Proxx game."""
        self.get_board_size()
        self.get_num_holes()
        self.create_board()
        self.place_holes()
        self.calculate_numbers()

        print("Initial board:")
        self.print_board()

        while True:
            try:
                row = int(input("Enter row (0 to {}): ".format(self.height - 1)))
                col = int(input("Enter column (0 to {}): ".format(self.width - 1)))

                if row < 0 or row >= self.height or col < 0 or col >= self.width:
                    print("Invalid move. Try again.")
                    continue

                if not self.reveal_cell(row, col):
                    print("Game over! You hit a hole.")
                    break

                if self.check_win():
                    print("Congratulations! You won the game!")
                    break

                self.print_board()
            except ValueError:
                print("Invalid input. Please enter a valid integer.")

        print("Final board:")
        self.print_board()